spec/struct.dd

```diff
@@ -555,10 +622,161 @@ $(H2 $(LNAME2 disable_default_construction, Disabling Default Struct Constructio
+            this(int i, int j)
+            {
+                this(i);
+                num = i + 1;  // assignment
+            }
+        }
+        ---
+        )
+
+        $(P If the field type has an $(LINK2 operatoroverloading.html#assignment, `opAssign`)
+        method, it will not be used for initialization.)
+
+        $(SPEC_RUNNABLE_EXAMPLE_COMPILE
+        ---
+        struct A
+        {
+            this(int n) {}
+            void $(CODE_HIGHLIGHT opAssign)(A rhs) {}
+        }
+        struct S
+        {
+            A val;
+            this(int i)
+            {
+                val = A(i);  // val is initialized to the value of A(i)
+                val = A(2);  // rewritten to val.opAssign(A(2))
+            }
+        }
+        ---
+        )
+
+        $(P If the field type is not mutable, multiple initialization will be rejected.)
+
+        $(SPEC_RUNNABLE_EXAMPLE_FAIL
+        ---
+        struct S
+        {
+            immutable int num;
+            this(int)
+            {
+                num = 1;  // OK
+                num = 2;  // Error: assignment to immutable
+            }
+        }
+        ---
+        )
+
+        $(P If the field is initialized on one path, it must be initialized on all paths.)
+        $(SPEC_RUNNABLE_EXAMPLE_FAIL
+        ---
+        struct S
+        {
+            immutable int num;
+            immutable int ber;
+            this(int i)
+            {
+                if (i)
+                    num = 3;   // initialization
+                else
+                    num = 4;   // initialization
+            }
+            this(long j)
+            {
+                j ? (num = 3) : (num = 4); // ok
+                j || (ber = 3);  // Error: intialized on only one path
+                j && (ber = 3);  // Error: intialized on only one path
+            }
+        }
+        ---
+        )
+
+        $(P A field initialization may not appear in a loop or after
+        a label.)
+
+        $(SPEC_RUNNABLE_EXAMPLE_FAIL
+        ---
+        struct S
+        {
+            immutable int num;
+            immutable string str;
+            this(int j)
+            {
+                foreach (i; 0..j)
+                {
+                    num = 1;    // Error: field initialization not allowed in loops
+                }
+                size_t i = 0;
+            Label:
+                str = "hello";  // Error: field initialization not allowed after labels
+                if (i++ < 2)
+                    goto Label;
+            }
+            this(int j, int k)
+            {
+                switch (j)
+                {
+                    case 1: ++j; break;
+                    default: break;
+                }
+                num = j;        // Error: `case` and `default` are also labels
+            }
+        }
+        ---
+        )
+
+        $(P If a field's type has disabled default construction, then it must be initialized
+        in the constructor.)
```

Comment:
Please mention that using the field before initialization results in an error (or if it doesn't then it is a bug)

Reply:
After this is merged : https://github.com/dlang/dmd/pull/8201 , the code in the original bug report provides a nice example which, in my opinion, should be present here.

Reply:
Again, I am documenting what is, not what should be. Just like we should not mix refactoring code with redesigning it.